### PR TITLE
BasePath support

### PR DESCRIPTION
**Pull request by rrooding**
Add basePath support as per #494

### CODE REVIEW

File: packages/serverless-component/__tests__/basepath.test.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    consoleWarnSpy.mockRestore();
+    fse.remove.mockRestore();
+    process.chdir(tmpCwd);
+  });
+
+  afterAll(cleanupFixtureDirectory(fixturePath));
+
+  describe("cloudfront", () => {
+    it("creates distribution", () => {
+      expect(mockCloudFront).toBeCalledWith({
```
Comment (danielcondemarin) [inline packages/serverless-component/__tests__/basepath.test.ts:73]: Could we change the assertion so we only test things affected by basePath? e.g. 

```
const cloudFrontCreateArgs = mockCreateCloudFront.mock.calls[0][0];
const pathPatterns = cloudFrontCreateArgs.origins[0].pathPatterns;
expect(Object.keys(pathPatterns)).toEqual("custom_base/_next/static/*", "custom_base/_next/data/*", ...)
```

File: packages/serverless-component/src/component.ts [hunk 2]
```diff
@@ -42,9 +43,26 @@ class NextjsComponent extends Component {
...
+  }
+
+  pathPattern(pattern: string, routesManifest: RoutesManifest): string {
+    const basePath = routesManifest.basePath;
+    if (basePath.length === 0) {
+      return pattern;
+    }
+
+    const basePathPattern =
+      basePath.charAt(0) === "/" ? basePath.slice(1) : basePath;
```
Comment (danielcondemarin) [inline packages/serverless-component/src/component.ts:57]: Out of curiosity why do we need this? Does next.js allow for base paths with and without `/`?

Comment (rrooding) [reply packages/serverless-component/src/component.ts]: It doesn't :) Fixed

File: README.md [hunk 3]
```diff
@@ -354,30 +354,30 @@ The fourth cache behaviour handles next API requests `api/*`.
...
+| bucketName               | `string`          | `null`                                                             | Custom bucket name where static assets are stored. By default is autogenerated.                                                                                                                                                                                                                                                                 |
+| bucketRegion             | `string`          | `null`                                                             | Region where you want to host your s3 bucket. Make sure this is geographically closer to the majority of your end users to reduce latency when CloudFront proxies a request to S3. Set this if you are experiencing 307 temporary redirects. See https://aws.amazon.com/premiumsupport/knowledge-center/s3-http-307-response/ for more details. |
```
Comment (danielcondemarin) [inline README.md:361]: The addition of "Set this if you are experiencing ... etc." should be deleted. That has been solved now.

File: packages/lambda-at-edge/src/default-handler.ts [hunk 4]
```diff
@@ -24,6 +26,11 @@ const addS3HostHeader = (
 
 const isDataRequest = (uri: string): boolean => uri.startsWith("/_next/data");
 
+const normaliseUri = (uri: string): string => {
+  if (basePath && uri.startsWith(basePath)) uri = uri.slice(basePath.length);
+  return uri === "/" || uri === "" ? "/index" : uri;
```
Comment (danielcondemarin) [inline packages/lambda-at-edge/src/default-handler.ts:31]: The `/` to `/index` normalisation isn't necessary anymore. I think you can just return `uri` directly after the basePath normalisation;

```suggestion
  return uri;
```

File: packages/serverless-component/src/component.ts [hunk 5]
```diff
@@ -42,9 +43,19 @@ class NextjsComponent extends Component {
     );
   }
 
+  readRoutesManifest(nextConfigPath: string): Promise<RoutesManifest> {
+    return readJSON(join(nextConfigPath, ".next/routes-manifest.json"));
+  }
+
+  pathPattern(pattern: string, routesManifest: RoutesManifest): string {
+    const basePath = routesManifest.basePath;
+    return basePath.length === 0 ? pattern : `${basePath.slice(1)}/${pattern}`;
```
Comment (danielcondemarin) [inline packages/serverless-component/src/component.ts:52]: Is this backward compatible with an older route manifest that doesn't include basePath? i.e. what if basePath is undefined?

File: packages/lambda-at-edge/src/default-handler.ts (hunk 4 above)
Comment (rrooding) [reply packages/lambda-at-edge/src/default-handler.ts:31]: Removing `uri === "/"` works. If I remove `uri === ""` then this specific test fails: https://github.com/serverless-nextjs/serverless-next.js/pull/510/files#diff-1e2c3ac8a5020d5e147d22a7d7df9029R49

File: packages/serverless-component/src/component.ts (hunk 5 above)
Comment (rrooding) [reply packages/serverless-component/src/component.ts]: Added an extra check for this :+1:

File: README.md (hunk 3 above)
Comment (rrooding) [reply README.md]: Fixed

**Comment by rrooding** [comment]
Not sure about the CI failures, seem to happen sometimes? Are they related to this PR? Locally everything is green.

File: packages/lambda-at-edge/src/default-handler.ts (hunk 4 above)
Comment (danielcondemarin) [reply packages/lambda-at-edge/src/default-handler.ts]: What if you do `return uri === "" ? "/" : uri;` ?

**Comment by danielcondemarin** [comment]
@rrooding I've pushed a fix for returning a 404 when the path doesn't start with the base path. It was matching both with and without the basepath. Also fixed `public/` files not using the `basepath/`

There seems to be an outstanding issue with trailing slashes. `exampledistrib.com/basepath/` is matched but `exampledistrib.com/basepath` isn't. Probably related to https://github.com/serverless-nextjs/serverless-next.js/issues/479 so I'll merge this PR anyway. 

Thanks for implementing 🙏

**Comment by danielcondemarin** [comment]
> Not sure about the CI failures, seem to happen sometimes? Are they related to this PR? Locally everything is green.

The CI issues seemed to be due to an attempt to cache node_modules https://github.com/serverless-nextjs/serverless-next.js/pull/510/commits/28396ffe6b5ec6937849501509d77bb64c7df89a. Sorted now

**Pull request merged**